lib/compiler/passes/merge-character-classes.js

```diff
@@ -0,0 +1,159 @@
+"use strict";
+
+const visitor = require("../visitor");
+
+function cloneOver(target, source) {
+  Object.keys(target).forEach(key => delete target[key]);
+  Object.keys(source).forEach(key => { target[key] = source[key]; });
+}
+
+// Clean up the parts array of a `class` node, by sorting,
+// then removing "contained" ranges, and merging overlapping
+// or adjacent ranges.
+function cleanParts(parts) {
+  // Sort parts on increasing start, and then decreasing end.
+  parts.sort((a, b) => {
+    const [aStart, aEnd] = Array.isArray(a) ? a : [a, a];
+    const [bStart, bEnd] = Array.isArray(b) ? b : [b, b];
+    if (aStart !== bStart) {
+      return aStart < bStart ? -1 : 1;
+    }
+    if (aEnd !== bEnd) {
+      return aEnd > bEnd ? -1 : 1;
+    }
+    return 0;
+  });
+
+  let prevStart = "";
+  let prevEnd = "";
+  for (let i = 0; i < parts.length; i++) {
+    const part = parts[i];
+    const [curStart, curEnd] = Array.isArray(part) ? part : [part, part];
+    if (curEnd <= prevEnd) {
+      // Current range is contained in previous range,
+      // so drop it.
+      parts.splice(i--, 1);
+      continue;
+    }
+    if (prevEnd.charCodeAt(0) + 1 >= curStart.charCodeAt(0)) {
+      // Current and previous ranges overlap, or are adjacent.
+      // Drop the current, and extend the previous range.
+      parts.splice(i--, 1);
+      parts[i] = [prevStart, prevEnd = curEnd];
+      continue;
+    }
+    prevStart = curStart;
+    prevEnd = curEnd;
+  }
+  return parts;
+}
+
+// Merges a choice character classes into a character class
+function mergeCharacterClasses(ast) {
+  // Build a map from rule names to rules for quick lookup of
+  // ref_fules.
+  const rules = Object.create(null);
+  ast.rules.forEach(rule => (rules[rule.name] = rule.expression));
+  // Keep a map of which rules have been processed, so that when
+  // we find a ref_rule, we can make sure its processed, before we
+  // try to use it.
+  const processedRules = Object.create(null);
+  const [asClass, merge] = [
+    node => {
+      if (node.type === "class" && !node.inverted) {
+        return node;
+      }
+      if (node.type === "literal" && node.value.length === 1) {
+        return {
+          type: "class",
+          parts: [node.value],
+          inverted: false,
+          ignoreCase: node.ignoreCase,
+          location: node.location,
+        };
+      }
+      if (node.type === "rule_ref") {
+        const ref = rules[node.name];
+        if (ref) {
+          if (!processedRules[node.name]) {
+            processedRules[node.name] = true;
+            merge(ref);
+          }
+          const cls = asClass(ref);
+          // Return a clone, not the original, because the value we return
+          // may get merged into.
+          return cls && {
+            type: cls.type,
+            parts: cls.parts.slice(),
+            inverted: cls.inverted,
+            ignoreCase: cls.ignoreCase,
+            location: node.location,
+          };
```

Comment:
Maybe it could be better to request cloned data from `asClass`, because we need to clone only `class` nodes, other nodes already created by us as new objects.

So probably
```javascript
return asClass(ref, true/* return a clone */);
```
here and `asClass(..., false)` otherwise else. Or even make clone at line 118
```javascript
prev = cls;
```
It seems the clone is really needed only there

Reply:
Makes sense.

While we're talking about clones though, it seems to be hard to produce generic clones in this project.

If I say `Object.assign({}, objectToClone)` eslint tells me to use `{ ...objectToClone }`, but if I say `{ ...objectToClone }`, it tells me `Parsing error: Unexpected token ...`.

I don't see why we shouldn't use `...` (it works with node14 and later, and will presumably get fixed by rollup)? Can I change the eslint settings?

Reply:
If you double-check that `...` gets replaced by the rollup step, submit a bug to https://github.com/peggyjs/peggyjs-eslint-config and I'll fix it.  There are a couple of other things I want to make sure are up-to-date while I'm in there.

Reply:
(it's actually the typescript step that is likely doing the replacement, which will be easier to check)